Pass URL to Browser::new(), delegate url checking logic to 3rd party

sadmansk opened the pull request:
<!-- Please describe your changes on the following line: -->
1. Move the logic of computing the initial url from `opts.rs` to `/ports/servo/main.rs`
2. Add a `ServoUrl` argument to `Browser::new`

Based on the requested changes by @paulrouget:
>We can read the pref in main() instead. shell.homepage would be used if the url is not passed as an argument. I'm trying to decouple the "app" logic and the "web engine" logic. I think it's up to the app to set the initial URL, and I'm not sure the initial url should be part of opts.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #15636 

<!-- Either: -->
- [ ] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16477)
<!-- Reviewable:end -->

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @cbrewster (or someone else) soon.

sadmansk commented on components/config/opts.rs:
```diff
@@ -657,11 +657,10 @@ pub fn from_cmdline_args(args: &[String]) -> ArgumentParsingResult {
         Some(&opt_match.free[0][..])
     } else {
-        homepage_pref.as_string()
+        None
```
Flag: what should the url_opt be set to if it can't be parsed from the cmdline?

paulrouget replied:
`None`.

paulrouget replied:
Also, Servo exits early if it can't find a URL. See https://github.com/sadmansk/servo/blob/95047114eeed77c4c9b310e6d4f04d3a3295a9e2/components/config/opts.rs#L678

This is not necessary anymore. Because the initial URL is now handled in main.rs, main.rs can do 2 things: either use `about:blank` or exist with the same error message (now sure which option is best).

sadmansk replied:
How would you like me to handle the case if it is null? Should I print the usage or simply assign `None`?

paulrouget replied:
Assign `None`.

sadmansk replied:
Forgive my newbiness, but I'm having some problem with assigning `None` on a matching statement. I couldn't search up exactly why it does not work. I was wondering if you could give me a pointer for the syntax.
```
error[E0308]: match arms have incompatible types
   --> /home/sadman/repos/servo/components/config/opts.rs:673:15
    |
673 |       let url = match url_opt {
    |  _______________^ starting here...
674 | |         Some(url_string) => {
675 | |             parse_url_or_filename(&cwd, url_string)
676 | |                 .unwrap_or_else(|()| args_fail("URL parsing failed"))
677 | |         },
678 | |         None => None,
679 | |     };
    | |_____^ ...ending here: expected struct `servo_url::ServoUrl`, found enum `std::option::Option`
    |
    = note: expected type `servo_url::ServoUrl`
               found type `std::option::Option<_>`
note: match arm with an incompatible type
   --> /home/sadman/repos/servo/components/config/opts.rs:678:17
    |
678 |         None => None,
    |                 ^^^^
```

paulrouget replied:
> Forgive my newbiness

I've been through this too ^^

> but I'm having some problem with assigning None on a matching statement. I couldn't search up exactly why it does not work. I was wondering if you could give me a pointer for the syntax.

Sure. The issue here is that the variable `url` is a ServoUrl. So the first branch of the `match` does return a ServoUrl. But the `None` branch returns a `Option<ServoUrl>`.

Before your PR, we had a `url_opt` (optional url: `Option<ServoUrl>`) that was transformed to url(`ServoUrl`). Because we didn't want url to be optional. Now we want. So we want to keep url optional.

So in this function, we don't want to `unwrap` the url (which mean moving from `Option<X>` to `X`), we want to keep it optional.

So this block of code your working on, was doing 2 things: 1) calling `parse_url_or_filename` to make sure it's a valid url and that a file path is transformed into a url, and 2) unwrap the url to enforce it (not make it optional).

Now we only want 1) to happen.

Let me know if you need more info. You can also ping me on irc (paul).

sadmansk replied:
First of all, thanks so much for giving me the time!
I do understand what we now want, I'm just having a bit of trouble understanding how to do it. Like, for instance, do I need to completely replace the match block and try unwrapping `url_opt` and save the value in `url_string`, and then only call `parse_url_or_file` on `url_string` if it isn't equal to either `None` or empty string?

paulrouget replied:
Sounds about right.

The `match` block should be replaced with something that would achieve this:

if `url_opt` is `None`, keep it `None`.
If `url_opt` is `Some(foo)`, `url_opt` should become `Some(parse_url_or_filename(foo).unwrap())`, but if the result of `parse_url_or_filename(foo)` is `Err()`, we should call `args_fail` (which will exit).

Achieving this should be easy with some combinaison of unwrap and map calls (methods of Option and Result types).

`parse_url_or_filename` return a `Result` that is more or less like an `Option`.

Make sure to understand:

* https://doc.rust-lang.org/std/option/
* https://doc.rust-lang.org/std/result/

This should be one or two lines of code. I let you try to understand what this should look like. Manipulating options is a core concept of Rust. It's worth the time understanding it.

Do not hesitate to ask more questions.

KiChjang replied:
The map function for Option should work very well here.

sadmansk commented on components/servo/lib.rs:
```diff
@@ -200,7 +203,7 @@ impl<Window> Browser<Window> where Window: WindowMethods + 'static {
         let (constellation_chan, sw_senders) = create_constellation(opts.user_agent.clone(),
                                                                     opts.config_dir.clone(),
-                                                                    opts.url.clone(),
+                                                                    target_url.clone(),
```
I'll fix the indentation issues in the next push.

sadmansk commented on components/servo/lib.rs:
```diff
@@ -322,9 +325,7 @@ fn create_constellation(user_agent: Cow<'static, str>,
-    if let Some(url) = url {
-        constellation_chan.send(ConstellationMsg::InitLoadUrl(url)).unwrap();
-    };
+    constellation_chan.send(ConstellationMsg::InitLoadUrl(url)).unwrap();
```
Flag: should we do a check here to see if url is `None`?

paulrouget replied:
It can't be `None` at this point.

paulrouget commented:
Thank you for taking care of this. I'll look at this PR early next week.

paulrouget commented on ports/servo/main.rs:
```diff
@@ -138,11 +140,14 @@ fn main() {
-        browser: Browser::new(window.clone()),
+        browser: Browser::new(window.clone(), opts::get().url.clone()
+            .unwrap_or(ServoUrl::parse(PREFS.get("shell.homepage").as_string().unwrap())
+            .unwrap()))
```
`opts::get().url` default value should be None, not Some("about:blank") ([see here](https://github.com/sadmansk/servo/blob/95047114eeed77c4c9b310e6d4f04d3a3295a9e2/components/config/opts.rs#L506)).

If `PREFS.get("shell.homepage")` doesn't exist, we should use `"about:blank"` (or exist early).

So you always have a valid URL.

sadmansk replied:
`about:blank` sounds like a good idea to me, I'll go with that.

cbrewster commented:
r? @paulrouget

cbrewster commented:
@bors-servo delegate=paulrouget

bors-servo commented:
:v: @paulrouget can now approve this pull request

paulrouget commented:
@sadmansk let me know if you have any question about my comments.

sadmansk commented:
@paulrouget just give me a few days, I have exams this week, I'll address your comments as soon as that's done :sweat_smile:

sadmansk committed:
Pass URL to Browser::new(), delegate url checking logic to third party

Use Option instead of ServoUrl for url in opts

Cleaner mapping from parse_url to url_opt

Add comment about url parsing error

Print reason for parsing error

Remove comment about warning

Add home url when openning new browser window in CEF

Fix merge error

sadmansk commented:
@paulrouget Made that change you requested. Please take another look and let me know if there was anything else you wanted me to change. Thanks.

KiChjang commented on ports/servo/main.rs:
```diff
@@ -143,7 +145,10 @@ fn main() {
+        browser: Browser::new(window.clone(),
+                opts::get().url.clone()
+                    .unwrap_or(ServoUrl::parse(PREFS.get("shell.homepage").as_string()
+                    .unwrap_or("about:blank")).unwrap()))
```
This line should get at least one more level of indentation to show that the method isn't operating at the same level as the previous line.

paulrouget commented on ports/servo/main.rs:
```diff
@@ -143,7 +145,10 @@ fn main() {
+        browser: Browser::new(window.clone(),
+                opts::get().url.clone()
+                    .unwrap_or(ServoUrl::parse(PREFS.get("shell.homepage").as_string()
+                        .unwrap_or("about:blank")).unwrap()))
```
Can you split that in a way that it's easier to understand?
Maybe move the url resolution above, and add a comment to explain the logic.

paulrouget reviewed (commented):
opts.rs' `from_cmdline_args` method needs some more work. `url_opt` can be `None`, and this is valid now as we fallback to the preference or about:blank in main.rs. See https://github.com/servo/servo/blob/733820583e41778b8e30faa4ba120ab0b0390073/components/config/opts.rs#L681 <- remove this

Also, the default url should be `None` I think: https://github.com/servo/servo/blob/733820583e41778b8e30faa4ba120ab0b0390073/components/config/opts.rs#L506

sadmansk commented:
@paulrouget I made the requested changes as I understood. Please let me know if I misunderstood anything. Also, if the parsing of the URL fails, i.e. `parse_url_or_filename()` returns an error, then I'm assigning `None` to `url` **without** any feedback. Should I print a message indicating that the parsing of the url failed?

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16757) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented on components/config/opts.rs:
```diff
@@ -667,13 +666,12 @@ pub fn from_cmdline_args(args: &[String]) -> ArgumentParsingResult {
+            match parse_url_or_filename(&cwd, url_string) {
+                Ok(v) => Some(v),
+                Err(_) => None,
+            }
```
You can transform an Result into an Option with parse_url_or_filename().ok().
See https://doc.rust-lang.org/std/result/enum.Result.html#method.ok

paulrouget commented on components/config/opts.rs:
```diff
@@ -795,7 +793,7 @@ pub fn from_cmdline_args(args: &[String]) -> ArgumentParsingResult {
     let opts = Opts {
         is_running_problem_test: is_running_problem_test,
-        url: Some(url),
+        url: url,
```
Keep `url_opt`:

```rust
url: url_opt
```

paulrouget commented on components/config/opts.rs:
```diff
@@ -667,13 +666,12 @@ pub fn from_cmdline_args(args: &[String]) -> ArgumentParsingResult {
-            args_fail("servo asks that you provide a URL")
-        }
+        None => None,
     };
```
I think you can combine all of this with `and_then`, something like this:

```rust
let url_opt = url_opt.and_then(|url_string| parse_url_or_filename(…).ok());
```

Also, rename `url` to `url_opt` (still an option at this point)

paulrouget reviewed (changes requested):
Address my comments, squash and rebase, and I think that should do it.

Thank you!

paulrouget commented:
> Also, if the parsing of the URL fails, i.e. parse_url_or_filename() returns an error, then I'm assigning None to url without any feedback. Should I print a message indicating that the parsing of the url failed?

Yes, maybe a warning.

sadmansk commented:
@paulrouget made the requested changes, sqaushed and rebased. Let me know if there is anything else.

paulrouget commented on components/config/opts.rs:
```diff
@@ -665,16 +664,12 @@ pub fn from_cmdline_args(args: &[String]) -> ArgumentParsingResult {
+                                   .or_else(|| {
+                                       println!("URL parsing failed.");
+                                       None
+                                   }));
```
You should use `warn!` instead of `println!`.

Also, `ok` transformed the `Result` into an `Option`. A `Result` is basically like an option, but with an error message instead of `None`.

So since we print a warning if it fails to parse, maybe we also want to display the error message that comes with `parse_url_or_filename`.

```rust
parse_url_or_filename(&cwd, url_string).or_else(|error| {
 warn!("URL parsing failed ({})", error);
 error
}).ok()
```

Does it make sense?

sadmansk replied:
Yeah it does, i'll push a change in a few mminutes.

sadmansk commented:
@paulrouget should be fine now, let me know if there's something else you want me changed.

paulrouget commented on components/config/opts.rs:
```diff
@@ -665,16 +664,12 @@ pub fn from_cmdline_args(args: &[String]) -> ArgumentParsingResult {
-        }
-    };
+    let url_opt = url_opt.and_then(|url_string| parse_url_or_filename(&cwd, url_string)
+                                   // If url parsing fails, print error and assign None
```
Update (or remove) this comment. The `warn!` is self-explanatory.

paulrouget commented:
Ok. It looks good now.

You also need to update `ports/cef`. You can build it with `./mach build-cef`.

sadmansk commented:
updated `ports/cef`. I had to set the home URL before making the call to the browser new. Is this okay?

paulrouget commented on ports/cef/browser.rs:
```diff
@@ -130,7 +132,9 @@ impl ServoCefBrowser {
-            let servo_browser = Browser::new(glutin_window.clone());
+            let home_url = ServoUrl::parse(PREFS.get("shell.homepage").as_string()
+                    .unwrap_or("about:blank")).unwrap();
+            let servo_browser = Browser::new(glutin_window.clone(), home_url);
```
If I'm not mistaken, ServoCefBrowser::new is called by `browser_host_create`, which comes with a url. So you should use it instead of use `about:blank`.

sadmansk replied:
Sorry a bit confused, should I assign `about:blank` or change the function signature of `ServoCefBrowser::new` so that it gets the url from `browser_host_create` and use that instead?

sadmansk replied:
@paulrouget Hey, still waiting on this clarification. Please let me know at your earliest convenience

paulrouget replied:
> should I change the function signature of ServoCefBrowser::new so that it gets the url from browser_host_create and use that instead?

Yes.

And then I don't think it's necessary to call `frame.set_url`: https://github.com/servo/servo/blob/d767e6046b5d970bc94818bc750afe5acccbdb31/ports/cef/browser.rs#L274

paulrouget commented on ports/cef/browser.rs:
```diff
@@ -9,6 +9,8 @@ use interfaces::{CefBrowser, CefBrowserHost, CefClient, CefFrame, CefRequestCont
 use interfaces::{cef_browser_t, cef_browser_host_t, cef_client_t, cef_frame_t};
 use interfaces::{cef_request_context_t};
 use servo::Browser;
+use servo::servo_config::prefs::PREFS;
```
I don't know much about CEF, but I think you should not use PREFS here. Just use `about:blank` directly.

paulrouget commented on ports/cef/browser.rs:
```diff
@@ -171,7 +175,9 @@ impl ServoCefBrowserExtensions for CefBrowser {
             window.set_browser(self.clone());
-            let servo_browser = Browser::new(window.clone());
+            let home_url = ServoUrl::parse(PREFS.get("shell.homepage").as_string()
+                    .unwrap_or("about:blank")).unwrap();
```
assign about:blank directly.

paulrouget commented:
@sadmansk anything I can do to help?

sadmansk commented:
@paulrouget Sorry about that, it's been a really busy week at work, so didn't get a chance to work on it. I will get to it as soon as I catch a breath

paulrouget commented:
No worries. Just wanted to make sure you were not stuck.

paulrouget commented:
@sadmansk if you want, we can land this as it is, and add the url to `ServoCefBrowser::new` in another PR.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17068) made this pull request unmergeable. Please resolve the merge conflicts.

sadmansk commented:
@paulrouget yeah ofcourse, that sounds like a good plan, i'll rebase tonight and ping you for the merge, and also start working on the cef browser stuff.

sadmansk commented:
@paulrouget merge is ready now, i'll work on the cef browser things on a different PR

paulrouget commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 708c561 has been approved by `paulrouget`

<!-- @bors-servo r=paulrouget 708c561ae698788c39d7d0ef7d796f349d7f218b -->

bors-servo commented:
:hourglass: Testing commit 708c561ae698788c39d7d0ef7d796f349d7f218b with merge 87140641a4f8636b431db41777d01302f8f6ad3d...

paulrouget commented:
CEF Follow up: #17273

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7441), [arm32](http://build.servo.org/builders/arm32/builds/7655), [arm64](http://build.servo.org/builders/arm64/builds/7695), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7784), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4540), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4563), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7766), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/251), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/250), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4469), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4536), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/246), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/249), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3011)
Approved by: paulrouget
Pushing 87140641a4f8636b431db41777d01302f8f6ad3d to master...
